src/ufs/mod.rs

```diff
@@ -769,17 +800,10 @@ impl<'a> Inode<'a> {
         let n = core::cmp::min(buf.len(), self.size() - off);
         let mut nread = 0;
         while nread < n {
-            let frag_off: usize = off % fragsize;
+            let frag_off = off % fragsize;
             let m = cmp::min(n - nread, fragsize - frag_off);
-            match self.bmap(off as u64)? {
-                Block::Hole => {
-                    buf[nread..nread + m].fill(0);
-                }
-                Block::Sd(bs) => {
-                    buf[nread..nread + m]
-                        .copy_from_slice(&bs[frag_off..frag_off + m]);
-                }
-            }
+            let block = self.bmap(off as u64)?;
+            block.read(frag_off, &mut buf[nread..nread + m]);
```

Comment:
this and `Block::read` are in conflict over who is responsible for sizing the read buffer appropriately. if `Block::Hole` was `Block::Hole(fragsize)` we could hold firm that `Block::read` can take a buffer of any size, and write up to `fragsize-frag_off` bytes into that buffer.

then we could pass only `&mut buf[nread..]`, return the number of bytes read from `Block::read`, and use that to bump `off` and `nread` instead of ever calculating `m` here. what do you think?

Reply:
I don't know that that's the case. The file size is in bytes, and may not be a multiple of the fragment size. The fragment size is fixed when the filesystem is created (it's a data in the superblock), but the file size is just whatever the file size is.

Reply:
Ok, we talked about this offline (or online?) and I'm convinced. PTAL.